log: Add validation interface logging

jkczyz opened the pull request:
Add logging of `CValidationInterface` callbacks using a new `VALIDATION` log flag (see #12994). A separate flag is desirable as the logging can be noisy and thus may need to be disabled without affecting other logging.

This could help debug issues where there may be race conditions at play, such as #12978.

jkczyz committed:
Add VALIDATION to BCLog::LogFlags

This flag is for logging from within CValidationInterface (see #12994).
A separate flag is desirable as the logging can be noisy and thus may
need to be disabled without affecting other logging.

jkczyz committed:
Add logging for CValidationInterface events

This could help debug issues where there may be race conditions at play,
such as #12978.

Fixes #12994.

ariard commented on src/validationinterface.cpp:
```diff
@@ -133,47 +137,62 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
-    m_internals->m_schedulerClient.AddToProcessQueue([pblock, pindex, pvtxConflicted, this] {
+    auto func = __func__;
+    m_internals->m_schedulerClient.AddToProcessQueue([func, pblock, pindex, pvtxConflicted, this] {
+        LogPrint(BCLog::VALIDATIONINTERFACE, "%s: block hash=%s\n", func, pblock->GetHash().ToString());
```
Thanks to pindex, maybe log the height too

jkczyz replied:
Added in 4bacb54.

ariard commented on src/validationinterface.cpp:
```diff
@@ -133,47 +137,62 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
-    m_internals->m_schedulerClient.AddToProcessQueue([locator, this] {
+    auto func = __func__;
+    m_internals->m_schedulerClient.AddToProcessQueue([func, locator, this] {
+        LogPrint(BCLog::VALIDATIONINTERFACE, "%s: block hash=%s\n", func, locator.IsNull() ? "null" : locator.vHave.front().ToString());
```
nit: maybe "top block hash %s bottom block hash"

ariard commented on src/validationinterface.cpp:
```diff
@@ -133,47 +137,62 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
-        m_internals->m_schedulerClient.AddToProcessQueue([ptx, this] {
+        auto func = __func__;
+        m_internals->m_schedulerClient.AddToProcessQueue([func, ptx, this] {
+            LogPrint(BCLog::VALIDATIONINTERFACE, "%s: tx hash=%s\n", func, ptx->GetHash().ToString());
```
You may add MemPoolRemovalReasonString somewhere, like TransactionErrorString in util/error.cp. IMO that's an interesting info to log

jkczyz replied:
Done in 504d009.

ariard reviewed (commented):
Tested validationinterface log flag with both cli option and logging RPC. Worked as expected.

Nice to have more logging around the validation callbacks, order of events may be so confusing!

hebasto commented on src/validationinterface.cpp:
```diff
@@ -5,9 +5,13 @@
+#include <chain.h>
+#include <logging.h>
 #include <primitives/block.h>
+#include <primitives/transaction.h>
```
From the [Developer Notes - Source code organization](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#source-code-organization):
> - Every `.cpp` and `.h` file should `#include` every header file it directly uses classes, functions or other definitions from, even if those headers are already included indirectly through other headers.
>    -  _Rationale_: Excluding headers because they are already indirectly included results in compilation failures when those indirect dependencies change. Furthermore, it obscures what the real code dependencies are.

ariard replied:
Ah thanks, didn't know it!

jnewbery commented:
Concept ACK

jkczyz commented on src/validationinterface.cpp:
```diff
@@ -133,47 +137,62 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
-    m_internals->m_schedulerClient.AddToProcessQueue([pindexNew, pindexFork, fInitialDownload, this] {
+    auto func = __func__;
+    m_internals->m_schedulerClient.AddToProcessQueue([func, pindexNew, pindexFork, fInitialDownload, this] {
+        if (pindexFork) LogPrint(BCLog::VALIDATIONINTERFACE, "%s: fork block hash=%s\n", func, pindexFork->GetBlockHash().ToString());
```
Good call! I had misinterpreted the frequency of this being non-null. Fixed in 4bacb54.

jkczyz commented:
Addressed all comments and add correct formatting for valid `CValidationState` in b303865.

Also updated `CValidationState` accessors in 3ff8755 to return strings by constant reference to avoid copying. Although this means the references are only valid for the lifetime of the object, which should be okay.

jnewbery commented on src/validationinterface.cpp:
```diff
@@ -128,52 +132,69 @@ void SyncWithValidationInterfaceQueue() {
-    m_internals->m_schedulerClient.AddToProcessQueue([ptx, this] {
+    auto func = __func__;
+    m_internals->m_schedulerClient.AddToProcessQueue([func, ptx, this] {
+        LogPrint(BCLog::VALIDATIONINTERFACE, "%s: tx hash=%s\n", func, ptx->GetHash().ToString());
```
nit: call this the 'txid' instead of 'tx hash', and also log the wtxid (use `GetWitnessHash()`)

jnewbery commented on src/txmempool.cpp:
```diff
@@ -1098,3 +1100,22 @@ void CTxMemPool::SetIsLoaded(bool loaded)
+{
+    switch (reason) {
+    case MemPoolRemovalReason::EXPIRY:
+        return "Expired from mempool";
```
nit: I think these strings could just be `EXPIRY`, `SIZELIMIT`, etc. These strings are for logs only, so will only be read by engineers.

Trying to make these strings more generally accessible obscures some of the nuance. For example `MemPoolRemovalReason::REORG` means that a transaction was removed during a re-org because it spent a coinbase transaction that is no longer mature, or because it is no longer final, or because mempool policy changed since it first entered the mempool so it can no longer be added or because the re-org was too large and we dropped transactions to avoid memory exhaustion.

jnewbery commented on src/validationinterface.cpp:
```diff
@@ -128,52 +132,69 @@ void SyncWithValidationInterfaceQueue() {
 
-    m_internals->m_schedulerClient.AddToProcessQueue([pindexNew, pindexFork, fInitialDownload, this] {
+    auto func = __func__;
+    m_internals->m_schedulerClient.AddToProcessQueue([func, pindexNew, pindexFork, fInitialDownload, this] {
```
For all of these async validationinterface callbacks, I think it'd be useful to log both when the event is enqueued and dequeued, with some kind of counter to correlate the two logs. Thoughts?

jkczyz replied:
The scheduler client guarantees these callbacks are executed serially in the order in which they were enqueued. Unless there is a need to correlate the enqueuing with the execution of `BlockChecked` and `NewPowValidBlock`, I'm hesitant to add more complexity to the code.

Is there a specific case where you think the additional logging would be useful? If so, would correlation using data from the log message be sufficient rather than adding a counter?

jnewbery replied:
I think this: https://github.com/bitcoin/bitcoin/issues/12978#issuecomment-381265760 would have been easier to spot with logging when the callbacks were enqueued, and it would have been easier to correlate those enqueue events with the thread that was running.

You're probably right that there's no need for a counter and that the data from the log message is enough to correlate the enqueue/dequeue events.

jkczyz replied:
Done. In order to avoid duplicating the logging code, I added some private template methods declared in the header and defined in the implementation file. The linker seems ok with this though let me know if there is a better way. Defining the helpers in the header would have required moving `MainSignalsInstance` into the header.

jkczyz replied:
FYI, I had to make some minor changes to placate the linter. It was giving the following failures:

```
src/validationinterface.cpp: Expected 0 argument(s) after format string but found 1 argument(s): LogPrint(BCLog::VALIDATIONINTERFACE, fmt, args...)
^---- failure generated from test/lint/lint-format-strings.sh
All calls to LogPrintf() and LogPrint() should be terminated with \n
src/validationinterface.cpp:    LogPrint(BCLog::VALIDATIONINTERFACE, fmt, args...);
^---- failure generated from test/lint/lint-logs.sh
```

https://travis-ci.org/bitcoin/bitcoin/jobs/599302990

mzumsande replied:
> In order to avoid duplicating the logging code, I added some private template methods declared in the header and defined in the implementation file. The linker seems ok with this(...)

This could lead to portability issues, see e.g. https://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file. I tried something similar recently in another PR, and the linker in one of the Travis runs did complain, so I am also interested to know if there are better solutions than moving stuff into the header.

jkczyz replied:
Yeah, that could very well be the case. That said, if we want to avoid argument evaluation when the logging category is not enabled, I would need to turn these into macros. See discussion starting in https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-543388763. Then I could remove those template method declarations from the header.

jnewbery replied:
This does indeed seem to have caused a build failure: https://travis-ci.org/bitcoin/bitcoin/jobs/600461225

jkczyz replied:
Indeed. It seems variadic arguments within lambda captures are not supported by that compilation environment. At least that's what I've gleaned from a similar issue (https://github.com/Ultimaker/CuraEngine/issues/426#issuecomment-269451925). However, this shouldn't be a problem if I change it into a macro.

jkczyz replied:
Turns out changing `EnqueueAndLogEvent` into a macro has some implications that are making this infeasible. Consider:

```c++
#define ENQUEUE_AND_LOG_EVENT(event, fmt, ...)                                             \
    do {                                                                                   \
        LOG_EVENT("Enqueuing " fmt, __VA_ARGS__);                                          \
        m_internals->m_schedulerClient.AddToProcessQueue([event, fmt, __VA_ARGS__, this] { \
            LOG_EVENT(fmt, __VA_ARGS__);                                                   \
            event();                                                                       \
        });                                                                                \
    } while (0)

#define LOG_EVENT(fmt, ...) \
    LogPrint(BCLog::VALIDATIONINTERFACE, fmt "\n", __VA_ARGS__);
```

The problem is that the capture list is expecting variables but the call sites are providing expressions. So unfortunately this won't compile. Presumably, if this did work, the expressions in the lambda would be evaluated thus defeating the purpose.

Unless there is a creative solution that I'm unaware of, the only alternative I can think of is to forgo the macro and duplicate the `LogPrint` for the enqueuing in each method. Thoughts?

jkczyz replied:
Actually, it looks like using a default capture instead of __VA_ARGS__ will do what I need!

jkczyz replied:
Implemented as a macro in 6ab424c.

jnewbery reviewed (commented):
I've partially reviewed and left a few comments inline. I intend to finish review and test this week.

jnewbery commented:
I did some light testing and this looks good. If you wanted to add automated tests, take a look at `assert_debug_log()` in the `TestNode` object. That allows you to test that specific strings have been written to the debug log.

> updated CValidationState accessors in 3ff8755 to return strings by constant reference to avoid copying.

Was this required for the PR? Those accessors aren't used very much, so performance isn't critical. I'd suggest not changing them in this PR.

jkczyz commented:
> I did some light testing and this looks good. If you wanted to add automated tests, take a look at `assert_debug_log()` in the `TestNode` object. That allows you to test that specific strings have been written to the debug log.
> 

I would be hesitant to add a functional test only to check for informational logging. Is there an existing a test where all these events are exercised? Not sure if it is worth adding regardless.

> > updated CValidationState accessors in [3ff8755](https://github.com/bitcoin/bitcoin/commit/3ff875575563bf88ea94ec63b892fe8237b5e0f7) to return strings by constant reference to avoid copying.
> 
> Was this required for the PR? Those accessors aren't used very much, so performance isn't critical. I'd suggest not changing them in this PR.

Yeah, commit 72c6e50 calls those methods in such a way that not returning references would have incurred excessive copies. Would it be better if the two commits were merged?

jkczyz commented:
@jnewbery I have some concerns about this additional logging. Since #14209 was merged, all log function arguments are evaluated regardless of whether logging is enabled for the given category. This was not the case prior to that PR since a macro was used for logging.

Given the heavy amount of logging added in this PR, should the function vs macro approach for logging be revisited? Typically, the use of macros in logging libraries is to avoid such argument evaluation if I understand correctly.

MarcoFalke commented:
> Since #14209 was merged, all log function arguments are evaluated regardless of whether logging is enabled for the given category.

Huh? They were passed to `strprintf` before and after that change, but `strprintf` was only called when logging was enabled. I don't see how that pull changed anything.

jkczyz commented:
> > Since #14209 was merged, all log function arguments are evaluated regardless of whether logging is enabled for the given category.
> 
> Huh? They were passed to `strprintf` before and after that change, but `strprintf` was only called when logging was enabled. I don't see how that pull changed anything.

By evaluation I mean at the call site of `LogPrint`. While this doesn't make a difference when passing variables, it does matter when the arguments are calls to other functions. Prior to #14209, the latter would be textually included when used and thus not evaluated if logging was not enabled for the given category.

MarcoFalke commented:
Ah, right

practicalswift commented on src/consensus/validation.h:
```diff
@@ -137,8 +137,8 @@ class CValidationState {
-    std::string GetRejectReason() const { return strRejectReason; }
-    std::string GetDebugMessage() const { return strDebugMessage; }
+    const std::string& GetRejectReason() const { return strRejectReason; }
+    const std::string& GetDebugMessage() const { return strDebugMessage; }
```
Why this change? :)

jkczyz replied:
See https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-543330180.

@jnewbery Regarding my previous comment, I'm also happy to remove this commit if that's preferred. Seems I had not read your comment closely enough about it not being performance critical. Though note my concerns about macro vs function logging.

practicalswift replied:
What is the object lifetime `strRejectReason` and `strDebugMessage`?

Are they guaranteed to outlive the callers of `GetRejectReason` or `GetDebugMessage`?

jkczyz replied:
Their lifetimes are that of the corresponding `CValidationState` as mentioned in https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-527575841. All current uses are guaranteed to outlive the callers as they either make a copy or are passed by const reference to other functions that do something similar. There's no such guarantee for future callers though.

practicalswift replied:
I don't think 6c730c3926d1871d8ce4c63eb6752808d7c3484a is a good idea: it introduces a very sharp edge in consensus critical code which is not offset by any measurable gain AFAICT.

We should always err on the side of caution when touching consensus critical code: if there is a trade-off between safety (current or future) and speed, then safety is generally much more important. In the cases where speed is critical for some reasons we should always measure before optimising :)

Somewhat related: https://www.youtube.com/watch?v=3MB2iiCkGxg&t=870

jkczyz replied:
Fair enough. I've removed 6c730c3.

hebasto commented on src/validationinterface.cpp:
```diff
@@ -133,47 +138,88 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
 }
+
+template <typename... Args>
+void CMainSignals::EnqueueAndLogEvent(std::function<void ()> event, const char* fmt, const Args&... args) {
```
Could `event` parameter be a const reference?

jkczyz replied:
No longer applicable as this is now a macros in 6ab424c.

hebasto reviewed (changes requested):
IMO, 6c730c3926d1871d8ce4c63eb6752808d7c3484a commit is outside the original PR goal.
Also `string` (i.e., `basic_string<char>`) offers move semantics; so it can be efficiently return by value.

jkczyz commented:
> IMO, [6c730c3](https://github.com/bitcoin/bitcoin/commit/6c730c3926d1871d8ce4c63eb6752808d7c3484a) commit is outside the original PR goal.
> Also `string` (i.e., `basic_string<char>`) offers move semantics; so it can be efficiently return by value.

I removed the commit. But I don't believe move semantics apply here as the function is returning a member variable.

jnewbery commented:
> I removed the commit. (https://github.com/bitcoin/bitcoin/commit/6c730c3926d1871d8ce4c63eb6752808d7c3484a)

Thanks!

I've tested this and it works great. I think we need to resolve this question: https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-543388763 before merging this PR. It seems reasonable to me to change logging back to using a macro for the reasons you suggest. Adding a `GetHash()` call to every validation interface call, even when logging is disabled, seems a shame. I'd be happy to review a PR that does that.

_Rename CMainSignals::MempoolEntryRemoved to TransactionRemovedFromMempool_ and _Format CValidationState properly in all cases_ could be pulled out into their own PR while we're waiting for that, although note that _Format CValidationState properly in all cases_ conflicts with https://github.com/bitcoin/bitcoin/pull/17004/commits/0053e16714323c1694c834fdca74f064a1a33529.

jkczyz commented:
> I've tested this and it works great. I think we need to resolve this question: [#16688 (comment)](https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-543388763) before merging this PR. It seems reasonable to me to change logging back to using a macro for the reasons you suggest. Adding a `GetHash()` call to every validation interface call, even when logging is disabled, seems a shame. I'd be happy to review a PR that does that.

@jnewbery Will do.

@MarcoFalke I'm not sure I follow the part about "unnecessary branches are not analyzed" in #14209. If I were to make `LogPrint` (but not `LogPrintf`) a macro again, will I need to define it conditionally based on `USE_COVERAGE` as was done in c8914b9?

MarcoFalke commented:
If you revert #14209, please make it as slim as possible (without the USE_COVERAGE guard, which mostly did only harm)

jnewbery commented:
> I would be hesitant to add a functional test only to check for informational logging. Is there an existing a test where all these events are exercised? Not sure if it is worth adding regardless.

Yeah, I agree that this isn't necessary.

jkczyz commented:
> _Rename CMainSignals::MempoolEntryRemoved to TransactionRemovedFromMempool_ and _Format CValidationState properly in all cases_ could be pulled out into their own PR while we're waiting for that, although note that _Format CValidationState properly in all cases_ conflicts with [0053e16](https://github.com/bitcoin/bitcoin/commit/0053e16714323c1694c834fdca74f064a1a33529).

I've rebased and resolved the conflict now that your change has been merged. The PR should be ready for any last comments. Also, will need reviews of #17218. :)

jnewbery commented:
It appears that the function name isn't being captured properly:

```
2019-11-01T15:06:44.834237Z [loadblk] Enqueuing BlockConnected: block hash=0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206 block height=0
2019-11-01T15:06:44.834255Z [loadblk] Enqueuing UpdatedBlockTip: new block hash=0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206 fork block hash=null (in IBD=1)
2019-11-01T15:06:44.834276Z [init] block tree size = 1
2019-11-01T15:06:44.834283Z [init] nBestHeight = 0
2019-11-01T15:06:44.834338Z [init] Bound to 127.0.0.1:11625
2019-11-01T15:06:44.834350Z [init] init message: Loading P2P addresses...
2019-11-01T15:06:44.834367Z [init] ERROR: DeserializeFileDB: Failed to open file /tmp/user/1000/bitcoin_func_test_if8trhon/node0/regtest/peers.dat
2019-11-01T15:06:44.834438Z [init] Invalid or missing peers.dat; recreating
2019-11-01T15:06:44.834758Z [scheduler] operator(): block hash=0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206 block height=0
2019-11-01T15:06:44.834919Z [scheduler] operator(): new block hash=0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206 fork block hash=null (in IBD=1)
```

Those last two lines should show `BlockConnected` and `UpdatedBlockTip`

jkczyz commented:
> It appears that the function name isn't being captured properly:

Fixed in 9eb373c.

jnewbery commented on src/validationinterface.cpp:
```diff
@@ -126,52 +131,90 @@ void SyncWithValidationInterfaceQueue() {
+// NOTE: The lambda captures all local variables by value.
+#define ENQUEUE_AND_LOG_EVENT(event, fmt, name, ...)           \
+    do {                                                       \
+        auto local_name = (name);                              \
```
At this point, it might just be easier to hard-code the function name into the log strings, or do you see a down-side to that?

jkczyz replied:
The downside is that the logging can go out of sync if the method is renamed (like in dd2855d) and that the format string is longer.

Note this implementation was used in an earlier version of the PR: https://github.com/bitcoin/bitcoin/pull/16688#pullrequestreview-278747465 (see second outdated comment). It was no longer needed when I added a helper method but needed again once I switched to a macro.

jnewbery commented on src/validationinterface.cpp:
```diff
@@ -126,52 +131,90 @@ void SyncWithValidationInterfaceQueue() {
+    ENQUEUE_AND_LOG_EVENT(event, "%s: new block hash=%s fork block hash=%s (in IBD=%d)", __func__,
+                          pindexNew->GetBlockHash().ToString(),
+                          pindexFork ? pindexFork->GetBlockHash().ToString() : "null",
+                          fInitialDownload);
```
nit: `fInitialDownload ? 'true' : 'false'`

jkczyz replied:
Done in f871e27 by changing `%d` to `%s`, as tinyformat does the conversion if using that specifier.

https://github.com/c42f/tinyformat#special-cases-for-p-c-and-s

jnewbery commented on src/validationinterface.cpp:
```diff
@@ -126,52 +131,90 @@ void SyncWithValidationInterfaceQueue() {
-    });
+    };
+    ENQUEUE_AND_LOG_EVENT(event, "%s: block hash=%s", __func__,
+                          pblock->GetHash().ToString());
```
Observation: This should be updated to also log the height when #15931 is merged (specifically commit https://github.com/bitcoin/bitcoin/pull/15931/commits/7d0f76a83e50328aaa8e7836b533ea0ea10d3c53)

jkczyz replied:
ACK. Happy to wait for that to be merged if you prefer.

jnewbery replied:
#15931 is merged

jkczyz replied:
Updated in 741cab0.

jnewbery reviewed (commented):
Tested ACK 9eb373c07ad78448ea78eee97d1ee9b9b409b4b9

One personal-style-preference nit, which you should feel free to ignore.

I've run several functional tests with this and manually tested it with a regtest node. It's really illuminating to see where the CValidationInterface callbacks are being enqueued and then called.

Really great stuff @jkczyz . Thanks for persisting with this!

jnewbery commented:
Tested ACK f871e272a1e4bd2b1c8b474416ee8eb065ef4e2c

ariard commented on src/txmempool.h:
```diff
@@ -356,6 +356,8 @@ enum class MemPoolRemovalReason {
     REPLACED,    //!< Removed for replacement
 };
 
+std::string MemPoolRemovalReasonString(const MemPoolRemovalReason reason);
```
nit: const?

jkczyz replied:
Do you mean remove the const from the parameter?

ariard replied:
Nevermind, that's not a class method, useless to put const in function signature

ariard commented on src/validationinterface.cpp:
```diff
@@ -126,52 +131,90 @@ void SyncWithValidationInterfaceQueue() {
     promise.get_future().wait();
 }
 
+// Use a macro instead of a function for conditional logging to prevent
```
Not a C++ expert, could you explain or point to why macro prevent to evaluate argument there? I mean aren't they replaced at preprocessing and we know flag only at runtime?

sipa replied:
compare the macro expansion:

```c++
if (debug_foo) {
    LogPrintf("Foobar %i", ExpensiveToComputeFunction());
}
```

with

```c++
LogPrint(BCLog::FOO, "Foobar %i", ExpensiveToComputeFunction());
```

In the first example, `ExpensiveToComputeFunction` is not invoked if `debug_foo` is false. In the second one it's invoked, but then disregarded by `LogPrint`.

ariard replied:
Thanks but got it, `LogPrint` has been changed to a macro in https://github.com/bitcoin/bitcoin/pull/17218, was still thinking it was a function...

That's said shouldn't we use the `LOG_PRINT` convention as for any other macro ?

jkczyz replied:
I'm using the convention in this PR but did not change it in the referenced PR for `LogPrint` since it was simply partially reverting a commit. It would have been a much larger change to update all the call sites, and that's something I'd be hesitant in undertaking without enough support for the change.

ariard reviewed (commented):
Tested with `-debug=validationinterface` set, print messages as expected. Holding my ACK on answer on the macro question.

ariard commented:
Tested ACK f871e27

jnewbery commented:
Rebased on master and added height logging to `BlockDisconnected` here: https://github.com/jnewbery/bitcoin/tree/pr16688.1 . Feel free to take it!

jnewbery commented:
Tested ACK 741cab0a7dcfe07860db8ad73678fbfde7eb6ada

Your latest push has different emails for the Author and Committer in the commits. I'm not sure if that was intentional or you want to change it back.

jkczyz commented:
> Tested ACK [741cab0](https://github.com/bitcoin/bitcoin/commit/741cab0a7dcfe07860db8ad73678fbfde7eb6ada)
> 
> Your latest push has different emails for the Author and Committer in the commits. I'm not sure if that was intentional or you want to change it back.

Updated and rebased.

jnewbery commented:
ACK cd6bf9970fdebc878cfc2d3b80bfc6cd4d39ffd7

ariard commented:
ACK cd6bf99. Change since last time is block height in `BlockDisconnected` log.

jkczyz commented on src/txmempool.cpp:
```diff
@@ -1098,3 +1100,22 @@ void CTxMemPool::SetIsLoaded(bool loaded)
+        return "CONFLICT";
+    case MemPoolRemovalReason::REPLACED:
+        return "REPLACED";
+    }
```
No longer needed since rebasing removed the need for this code.

jkczyz commented:
Rebased and resolved merge conflicts. The commits for formatting `MemPoolRemovalReason` and renaming `CMainSignals::MempoolEntryRemoved` have been removed since #14384 made them obsolete. But just realized @jnewbery asked to have the reason added in this PR if the #14384 was merged first.

jnewbery commented:
> @jnewbery asked to have the reason added in this PR

It's not essential but I think logging the reason is definitely a nice-to-have. None of the clients need access to the reason (yet) so it could just be logged and dropped in the validationinterface.

jkczyz commented:
> It's not essential but I think logging the reason is definitely a nice-to-have. None of the clients need access to the reason (yet) so it could just be logged and dropped in the validationinterface.

Done in 34d6486.

jnewbery commented:
Tested ACK 34d6486de309d24d040a900fe3492582dc6bbb45

jnewbery commented:
I'd love to see this merged soon. Once you've had validationinterface logging for test debugging, you don't want to go back!

jkczyz commented:
The linter says 34d6486 introduced a circular dependency:

```
A new circular dependency in the form of "txmempool -> validationinterface -> txmempool" appears to have been introduced.
```

Suggestions on where `MemPoolRemovalReasonString` should live?

jnewbery commented:
> To break this, I'd have to move both MemPoolRemovalReason and MemPoolRemovalReasonString elsewhere. Thoughts?

Just remove that commit. No need to make this PR more complex just to log the removal reason. It can be done in a follow-up.

jnewbery commented:
Tested ACK 0cadb123b

ariard commented:
ACK 0cadb12, only changes are dropping commits/removing args following #14384 merge

hebasto commented:
Concept ACK. Will test.

hebasto commented:
There are some code organization concerns.

I believe that headers and forward class declaration should not be touched in the "_Format CValidationState properly in all cases_" commit (ee4d66a38c5c09161f8a3e63047db6b6908687f4).

Instead, `<consensus/validation.h>` should be included to `validationinterface.cpp` in the "_Add logging for CValidationInterface events_" commit (0cadb123bad37a2c224834a3d98bed474c661c21).

hebasto reviewed (changes requested):
Tested on Linux Mint 19.3. It works as expected.

The only suggestion: https://github.com/bitcoin/bitcoin/pull/16688#issuecomment-568243356

ryanofsky commented on src/logging.cpp:
```diff
@@ -162,6 +162,7 @@ const CLogCategoryDesc LogCategories[] =
     {BCLog::COINDB, "coindb"},
     {BCLog::QT, "qt"},
     {BCLog::LEVELDB, "leveldb"},
+    {BCLog::VALIDATIONINTERFACE, "validationinterface"},
```
In commit "Add VALIDATIONINTERFACE to BCLog::LogFlags" (389d423a252e43d0b08871f8ae5b66a84a8a217c)

Would change names to `VALIDATION, "validation"`. These are just validation events. Fact that they are printed in the validationinterface source file instead of the validation one is an implementation detail that could change and would be arcane to someone just trying to configure logging categories.

ryanofsky commented on src/util/validation.h:
```diff
@@ -6,9 +6,9 @@
-#include <string>
+#include <consensus/validation.h>
 
-class ValidationState;
```
In commit "Format CValidationState properly in all cases" (ee4d66a38c5c09161f8a3e63047db6b6908687f4)

Replacing forward declaration with a full include seems unnecessary here, just making the diff bigger and perhaps unnecessarily slowing down compilation.

ryanofsky commented on src/util/validation.cpp:
```diff
@@ -5,15 +5,20 @@
+    const std::string debug_message = state.GetDebugMessage();
+    if (!debug_message.empty()) {
+        return strprintf("%s, %s", state.GetRejectReason(), debug_message);
+    }
```
In commit "Format CValidationState properly in all cases" (ee4d66a38c5c09161f8a3e63047db6b6908687f4)

I'm a little confused by all the changes in this commit. As far as I can tell the only change that affects behavior is adding `if (state.IsValid()) return "Valid";` at the top of this function, and everything else is a rewrite / style cleanup. Not a big deal, but in the future it might be good to fully motivate changes in the commit description or do style changes in separate commits.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -110,52 +114,88 @@ void SyncWithValidationInterfaceQueue() {
-void CMainSignals::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock, const CBlockIndex* pindex)
-{
-    m_internals->m_schedulerClient.AddToProcessQueue([pblock, pindex, this] {
+void CMainSignals::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock, const CBlockIndex* pindex) {
```
In commit "Add logging for CValidationInterface events" (0cadb123bad37a2c224834a3d98bed474c661c21)

Whitespace is changing unnecessarily (and format is nonstandard, brace should go on new line)

jkczyz replied:
I was largely trying to avoid style changes and conform with the local conventions in the file. However, in the interim the brace was moved in https://github.com/bitcoin/bitcoin/commit/10b4729e33f76092bd8cfa06d1a5e0a066436f76#diff-e8d9e22d9683f73a9fb8399be0dab640R159 to use the standard conventions.

Not sure what is preferred, but I don't mind changing it back to reduce the diff size.

ryanofsky reviewed (approved):
Code review ACK 0cadb123bad37a2c224834a3d98bed474c661c21. Nice change! I left some suggestions, but feel free to ignore them.

jkczyz committed:
Format CValidationState properly in all cases

FormatStateMessage does not properly handle the case where
CValidationState::IsValid() returns true. Use "Valid" for the state in
this case.

jkczyz committed:
Refactor FormatStateMessage for clarity

All cases of CValidationState were condensed into one strprintf call.
This is no longer suitable as more cases are added (e.g., IsValid).

jkczyz commented:
@hebasto @ryanofsky Thanks for the review! I've addressed all comments.

jnewbery commented:
ACK f9abf4ab6d3d3e4d4b7e90723020b5422a141a6f

hebasto commented:
ACK f9abf4ab6d3d3e4d4b7e90723020b5422a141a6f

ariard commented:
ACK f9abf4a, only changes since 0cadb12 are replacing `VALIDATIONINTERFACE` by `VALIDATION` and avoiding a forward declaration with a new include

ryanofsky reviewed (approved):
Code review ACK f9abf4ab6d3d3e4d4b7e90723020b5422a141a6f. Just suggested changes since last review (thanks!)

jonatack reviewed (commented):
ACK e7f84503571c171a7e6728cd2d77dd4103bd7a6f  posthumous code review and manual tests.

Usage listed here for visitor convenience (the PR description was not updated from validationinterface to validation):

- `bitcoind -help | grep -B6 ", validation"`
```
  -debug=<category>
       Output debugging information (default: -nodebug, supplying <category> is
       optional). If <category> is not supplied or if <category> = 1,
       output all debugging information. <category> can be: net, tor,
       mempool, http, bench, zmq, walletdb, rpc, estimatefee, addrman,
       selectcoins, reindex, cmpctblock, rand, prune, proxy, mempoolrej,
       libevent, coindb, qt, leveldb, validation.
```
- `bitcoind -debug=validation` (launch bitcoind with validation logging)
- `bitcoin-cli help logging`
- `bitcoin-cli logging [] '["validation"]'` (toggle validation logging off)
- `bitcoin-cli logging '["validation"]'` (toggle validation logging on)
